### PR TITLE
add data_source_huaweicloud_gaussdb_mysql_instances

### LINKED ISSUE #853: [Feature Request] data sources to get all available gaussdb mysql instances
Hi there,

Now, the huaweicloud provider already has data_source_huaweicloud_gaussdb_mysql_instance to get an available instance.
It's very helpful if there could be a data source to get all available instances.

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
fixes #853 
Test result:
```bash
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccGaussdbMysqlInstancesDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccGaussdbMysqlInstancesDataSource_basic -timeout 360m -parallel 4
=== RUN   TestAccGaussdbMysqlInstancesDataSource_basic
=== PAUSE TestAccGaussdbMysqlInstancesDataSource_basic
=== CONT  TestAccGaussdbMysqlInstancesDataSource_basic
--- PASS: TestAccGaussdbMysqlInstancesDataSource_basic (730.98s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       731.020s
```

### CODE REVIEW

File: huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("GaussDB mysql instances data source ID not set ")
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckGaussdbMysqlInstancesDataLen(n string) resource.TestCheckFunc {
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances_test.go]: seems this function is never called.

BTW, this check funcction is not recommended, you can try `resource.TestCheckResourceAttr("data.huaweicloud_gaussdb_mysql_instances.test", "instances.#", "1"),`

File: huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+			"Please change your search criteria and try again.")
+	}
+
+	var instancesToSet []map[string]interface{}
+	var instancesIds []string
+
+	for _, instanceInAll := range allInstances.Instances {
+		instanceID := instanceInAll.Id
+		instancesIds = append(instancesIds, instanceID)
+		instance, err := instances.Get(client, instanceID).Extract()
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances.go]: why we get the DB details again?

File: huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances_test.go (hunk 1 above)
Comment (Jason-Zhang9309) [reply huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances_test.go]: fixed

File: huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances.go (hunk 2 above)
Comment (Jason-Zhang9309) [reply huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances.go]: DB detail returns more information than list instances, such as nodes.
You can compare the API docs.

File: huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances_test.go (hunk 1 above)
Comment (Jason-Zhang9309) [reply huaweicloud/data_source_huaweicloud_gaussdb_mysql_instances_test.go]: new test result:
```bash
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccGaussdbMysqlInstancesDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccGaussdbMysqlInstancesDataSource_basic -timeout 360m -parallel 4
=== RUN   TestAccGaussdbMysqlInstancesDataSource_basic
=== PAUSE TestAccGaussdbMysqlInstancesDataSource_basic
=== CONT  TestAccGaussdbMysqlInstancesDataSource_basic
--- PASS: TestAccGaussdbMysqlInstancesDataSource_basic (813.94s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       813.983s
```

**Comment by Jason-Zhang9309** [comment]
use getting detail only in necessary parts, make the code logic clearer

new test result:
```bash
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccGaussdbMysqlInstancesDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccGaussdbMysqlInstancesDataSource_basic -timeout 360m -parallel 4
=== RUN   TestAccGaussdbMysqlInstancesDataSource_basic
=== PAUSE TestAccGaussdbMysqlInstancesDataSource_basic
=== CONT  TestAccGaussdbMysqlInstancesDataSource_basic
--- PASS: TestAccGaussdbMysqlInstancesDataSource_basic (797.28s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       797.328s
```

**Pull request merged**